class/practice6.1/include/Dispatcher.hpp

```diff
@@ -0,0 +1,90 @@
+#pragma once
+#include <string>
+#include <vector>
+#include <mutex>
+#include <functional>
+#include <random>
+#include <thread>
+
+struct GreenMessage {
+    std::string message;
+    int counter;
+};
+
+struct BlueMessage {
+    double val1;
+    double val2;
+};
+
+struct OrangeMessage {
+    std::string message1;
+    std::string message2;
+    int counter;
+    double val;
+};
+class MessageDispatcher {
+public:
+    void subscribeGreen(const std::function<void(const GreenMessage&)>& subscriber) {
+        std::lock_guard<std::mutex> lock(mutex);
+        greenSubscribers.push_back(subscriber);
+    }
+
+    void subscribeBlue(const std::function<void(const BlueMessage&)>& subscriber) {
+        std::lock_guard<std::mutex> lock(mutex);
+        blueSubscribers.push_back(subscriber);
+    }
+
+    void subscribeOrange(const std::function<void(const OrangeMessage&)>& subscriber) {
+        std::lock_guard<std::mutex> lock(mutex);
+        orangeSubscribers.push_back(subscriber);
+    }
+
+    void publishGreen(const GreenMessage& message) {
+        std::lock_guard<std::mutex> lock(mutex);
+        for (const auto& subscriber : greenSubscribers) {
+            subscriber(message);
```

Comment:
it may be  a problem to process message just here as in case of complex callback we may hang a thread of publisher while caller may expect instant publish.
Therefore usually we should have a separate thread that will dispatch messages and will call handlers